### PR TITLE
Offset for Every Step

### PR DESCRIPTION
**Pull request by swalker2m**
Pulls the `StepConfig.Science` telescope configuration (offset and guiding) up to `Step` itself.  This information was already part of `Science` steps but will now be included in all steps.  The telescope configuration is separated into its own `TelescopeConfig`.

There is more discussion of this at the end of ShortCut 3374, starting here: https://app.shortcut.com/lucuma/story/3374/update-gmos-sequence-generator#activity-3731

### CODE REVIEW

File: modules/core/shared/src/main/scala/lucuma/core/model/sequence/StepConfig.scala [hunk 1]
```diff
@@ -33,24 +32,18 @@ object StepConfig {
     lamp:      Gcal.Lamp,
     filter:    GcalFilter,
     diffuser:  GcalDiffuser,
-    shutter:   GcalShutter
-  ) extends StepConfig(StepType.Gcal) {
+    shutter:   GcalShutter,
+    telescope: TelescopeConfig
```
Comment (swalker2m) [inline modules/core/shared/src/main/scala/lucuma/core/model/sequence/StepConfig.scala:36]: `TelescopeConfig` added here.

File: modules/core/shared/src/main/scala/lucuma/core/model/sequence/StepConfig.scala [hunk 2]
```diff
@@ -119,45 +110,43 @@ object StepConfig {
...
+    val telescope: Lens[Gcal, TelescopeConfig] =
+      Focus[Gcal](_.telescope)
+  end Gcal
 
-    /** @group Optics */
-    val guiding: Lens[Science, StepGuideState] =
-      Focus[Science](_.guiding)
-  }
+  final case class Science(
+    telescope: TelescopeConfig
```
Comment (swalker2m) [inline modules/core/shared/src/main/scala/lucuma/core/model/sequence/StepConfig.scala:120]: Switch to embed `TelescopeConfig` here.  I anticipate changing the GraphQL API accordingly.

File: modules/core/shared/src/main/scala/lucuma/core/model/sequence/StepConfig.scala [hunk 3]
```diff
@@ -119,45 +110,43 @@ object StepConfig {
...
-  final case class SmartGcal(smartGcalType: SmartGcalType) extends StepConfig(StepType.SmartGcal)
+  object Science:
+    /** @group Optics */
+    val telescope: Lens[Science, TelescopeConfig] =
+      Focus[Science](_.telescope)
 
-  object SmartGcal {
+  final case class SmartGcal(
+    smartGcalType: SmartGcalType,
+    telescope:     TelescopeConfig
```
Comment (swalker2m) [inline modules/core/shared/src/main/scala/lucuma/core/model/sequence/StepConfig.scala:130]: Add `TelescopeConfig` here.

File: modules/core/shared/src/main/scala/lucuma/core/model/sequence/TelescopeConfig.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import cats.derived.*
+import lucuma.core.enums.StepGuideState
+import lucuma.core.math.Offset
+import monocle.Focus
+import monocle.Lens
+
+case class TelescopeConfig(
+  offset:  Offset,
+  guiding: StepGuideState
+) derives Order
```
Comment (swalker2m) [inline modules/core/shared/src/main/scala/lucuma/core/model/sequence/TelescopeConfig.scala:16]: Previously just part of `StepConfig.Science`.

**Review by swalker2m — COMMENTED** [review]
Here I've marked the important changes so you don't have to squint through the noise.

File: modules/core/shared/src/main/scala/lucuma/core/model/sequence/Step.scala [hunk 5]
```diff
@@ -22,17 +24,15 @@ case class Step[+D](
   id:                Step.Id,
   instrumentConfig:  D,
   stepConfig:        StepConfig,
+  telescopeConfig:   TelescopeConfig,
```
Comment (swalker2m) [inline modules/core/shared/src/main/scala/lucuma/core/model/sequence/Step.scala:27]: `TelescopeConfig` added to all steps here.

File: modules/core/shared/src/main/scala/lucuma/core/model/sequence/StepConfig.scala [hunk 6]
```diff
@@ -119,43 +109,25 @@ object StepConfig {
     /** @group Optics */
     val shutter: Lens[Gcal, GcalShutter] =
       Focus[Gcal](_.shutter)
-  }
+  end Gcal
 
-  final case class Science(offset: Offset, guiding: StepGuideState) extends StepConfig(StepType.Science)
+  case object Science extends StepConfig(StepType.Science)
```
Comment (swalker2m) [inline modules/core/shared/src/main/scala/lucuma/core/model/sequence/StepConfig.scala:114]: The telescope config is moved out of `StepConfig.Science` here.

**Review by toddburnside — APPROVED** [review]
LGTM

File: modules/core/shared/src/main/scala/lucuma/core/model/sequence/StepConfig.scala [hunk 7]
```diff
@@ -18,12 +18,11 @@ import monocle.std.either.*
 
 import scala.collection.immutable.SortedSet
 
-sealed abstract class StepConfig(val stepType: StepType) {
+sealed abstract class StepConfig(val stepType: StepType):
```
Comment (cquiroz) [inline modules/core/shared/src/main/scala/lucuma/core/model/sequence/StepConfig.scala:21]: Embracing significant whitespce?

**Review by cquiroz — APPROVED** [review]
LGTM

**Pull request merged**